File: apps/web/__tests__/lib/call-analytics.test.ts
/**
 * Pure web call-telemetry accumulator — the SSOT for the `call:analytics`
 * payload the web reports at hangup. Parité iOS `emitCallAnalyticsSnapshot` /
 * Android `CallAnalytics`: the web was the ONE client that never emitted this
 * telemetry (prod 2026-07-12: 100% of analytics rows were iOS), leaving the
 * reliability dashboard blind to web calls. Pure + total so every branch is
 * unit-tested without React or a socket.
 */

import {
  createCallAnalytics,
  markConnected,
  markReconnecting,
  addQualitySample,
  buildAnalyticsPayload,
} from '@/lib/call-analytics';

describe('call-analytics accumulator', () => {
  it('setupTimeMs is the connect delay for a call that connected', () => {
    let acc = createCallAnalytics(1_000);
    acc = markConnected(acc, 4_200);

    const payload = buildAnalyticsPayload(acc, {
      callId: 'c1', nowMs: 60_000, isVideo: false, endReason: 'local',
    });

    expect(payload.setupTimeMs).toBe(3_200);
  });

  it('setupTimeMs is the -1 « never connected » sentinel when the call never connected', () => {
    const acc = createCallAnalytics(1_000);

    const payload = buildAnalyticsPayload(acc, {
      callId: 'c1', nowMs: 5_000, isVideo: false, endReason: 'missed',
    });

    expect(payload.setupTimeMs).toBe(-1);
  });

  it('only the FIRST connect anchors setupTimeMs (a reconnect never re-anchors)', () => {
    let acc = createCallAnalytics(1_000);
    acc = markConnected(acc, 3_000);
    acc = markConnected(acc, 50_000);

    const payload = buildAnalyticsPayload(acc, {
      callId: 'c1', nowMs: 60_000, isVideo: false, endReason: 'local',
    });

    expect(payload.setupTimeMs).toBe(2_000);
  });

  it('counts reconnections', () => {
    let acc = createCallAnalytics(0);
    acc = markConnected(acc, 1_000);
    acc = markReconnecting(acc);
    acc = markReconnecting(acc);

    const payload = buildAnalyticsPayload(acc, {
      callId: 'c1', nowMs: 30_000, isVideo: false, endReason: 'local',
    });

    expect(payload.reconnectionCount).toBe(2);
  });

  it('averages RTT and packet loss over samples and keeps the max packet loss', () => {
    let acc = createCallAnalytics(0);
    acc = addQualitySample(acc, { level: 'good', rtt: 100, packetLoss: 1 });
    acc = addQualitySample(acc, { level: 'fair', rtt: 300, packetLoss: 5 });

    const payload = buildAnalyticsPayload(acc, {
      callId: 'c1', nowMs: 10_000, isVideo: true, endReason: 'local',
    });

    expect(payload.averageRtt).toBe(200);
    expect(payload.averagePacketLoss).toBe(3);
    expect(payload.maxPacketLoss).toBe(5);
  });

  it('reports zeroed quality metrics when no sample was taken (never connected)', () => {
    const acc = createCallAnalytics(0);

    const payload = buildAnalyticsPayload(acc, {
      callId: 'c1', nowMs: 5_000, isVideo: false, endReason: 'missed',
    });

    expect(payload.averageRtt).toBe(0);
    expect(payload.averagePacketLoss).toBe(0);
    expect(payload.maxPacketLoss).toBe(0);
  });

  it('builds a normalized quality distribution (fractions summing ~1)', () => {
    let acc = createCallAnalytics(0);
    acc = addQualitySample(acc, { level: 'excellent', rtt: 20, packetLoss: 0 });
    acc = addQualitySample(acc, { level: 'excellent', rtt: 25, packetLoss: 0 });
    acc = addQualitySample(acc, { level: 'poor', rtt: 500, packetLoss: 10 });

    const payload = buildAnalyticsPayload(acc, {
      callId: 'c1', nowMs: 30_000, isVideo: false, endReason: 'local',
    });

    expect(payload.qualityDistribution.excellent).toBeCloseTo(2 / 3, 5);
    expect(payload.qualityDistribution.poor).toBeCloseTo(1 / 3, 5);
    expect(payload.qualityDistribution.good).toBe(0);
    expect(payload.qualityDistribution.fair).toBe(0);
  });

  it('an empty distribution (no samples) is all zeros', () => {
    const acc = createCallAnalytics(0);
    const payload = buildAnalyticsPayload(acc, {
      callId: 'c1', nowMs: 1_000, isVideo: false, endReason: 'missed',
    });
    expect(payload.qualityDistribution).toEqual({ excellent: 0, good: 0, fair: 0, poor: 0 });
  });

  it('carries the durationSeconds derived from start→now, callId, platform=web and honest defaults', () => {
    let acc = createCallAnalytics(2_000);
    acc = markConnected(acc, 3_000);

    const payload = buildAnalyticsPayload(acc, {
      callId: 'call-xyz', nowMs: 62_000, isVideo: true, endReason: 'remote',
    });

    expect(payload.callId).toBe('call-xyz');
    expect(payload.durationSeconds).toBe(60);
    expect(payload.isVideo).toBe(true);
    expect(payload.endReason).toBe('remote');
    expect(payload.platform).toBe('web');
    // Honest defaults for signals the web doesn't track.
    expect(payload.networkTransitions).toBe(0);
    expect(payload.filtersUsed).toBe(false);
    expect(payload.effectsUsed).toEqual([]);
  });
});

File: apps/android/core/model/src/test/kotlin/me/meeshy/sdk/model/call/CallAnalyticsTest.kt
package me.meeshy.sdk.model.call

import com.google.common.truth.Truth.assertThat
import org.junit.Test

/**
 * Behavioural spec for the pure once-per-call analytics accumulator behind
 * `call:analytics` (iOS parity: `emitCallAnalyticsSnapshot`). [CallAnalytics.fields]
 * is the single tested SSOT of the wire payload — the gateway validates every
 * field with `socketCallAnalyticsSchema`, so each key asserted here is a key the
 * server would otherwise reject at runtime.
 */
class CallAnalyticsTest {

    private fun sample(rtt: Double, loss: Double) = CallQualitySample(rttMs = rtt, packetLoss = loss)

    private fun fieldsOf(
        analytics: CallAnalytics,
        durationSeconds: Long = 0,
        isVideo: Boolean = false,
        endReason: CallEndReason = CallEndReason.Local,
    ): Map<String, Any> = analytics.fields(durationSeconds, isVideo, endReason, deviceModel = "Pixel 8")

    // --- anchors --------------------------------------------------------------

    @Test
    fun `setupTimeMs is the start-to-first-connect delta`() {
        val analytics = CallAnalytics(startedAtMs = 5_000).connected(nowMs = 8_200)

        assertThat(fieldsOf(analytics)["setupTimeMs"]).isEqualTo(3_200L)
    }

    @Test
    fun `connected is idempotent so a reconnect never rewrites the anchor`() {
        val analytics = CallAnalytics(startedAtMs = 5_000).connected(8_000).connected(60_000)

        assertThat(fieldsOf(analytics)["setupTimeMs"]).isEqualTo(3_000L)
    }

    @Test
    fun `a call that never connected reports the -1 sentinel`() {
        val analytics = CallAnalytics(startedAtMs = 5_000)

        assertThat(fieldsOf(analytics)["setupTimeMs"]).isEqualTo(-1L)
    }

    @Test
    fun `negotiationTimeMs is omitted — no anchor on Android, the schema allows absence`() {
        assertThat(fieldsOf(CallAnalytics(startedAtMs = 0))).doesNotContainKey("negotiationTimeMs")
    }

    // --- reconnects -----------------------------------------------------------

    @Test
    fun `each reconnect cycle increments the count`() {
        val analytics = CallAnalytics(startedAtMs = 0).reconnecting().reconnecting()

        assertThat(fieldsOf(analytics)["reconnectionCount"]).isEqualTo(2)
    }

    // --- quality aggregation ----------------------------------------------------

    @Test
    fun `samples aggregate into average and max`() {
        val analytics = CallAnalytics(startedAtMs = 0)
            .plusSample(sample(rtt = 100.0, loss = 1.0))
            .plusSample(sample(rtt = 300.0, loss = 7.0))

        val fields = fieldsOf(analytics)
        assertThat(fields["averageRtt"]).isEqualTo(200.0)
        assertThat(fields["averagePacketLoss"]).isEqualTo(4.0)
        assertThat(fields["maxPacketLoss"]).isEqualTo(7.0)
    }

    @Test
    fun `no samples report zero averages rather than NaN`() {
        val fields = fieldsOf(CallAnalytics(startedAtMs = 0))

        assertThat(fields["averageRtt"]).isEqualTo(0.0)
        assertThat(fields["averagePacketLoss"]).isEqualTo(0.0)
        assertThat(fields["maxPacketLoss"]).isEqualTo(0.0)
    }

    @Test
    fun `quality distribution carries every tier as a fraction of samples`() {
        // rtt 50/loss 0 → EXCELLENT ; rtt 350/loss 6 → POOR (via the tier ladder).
        val analytics = CallAnalytics(startedAtMs = 0)
            .plusSample(sample(rtt = 50.0, loss = 0.0))
            .plusSample(sample(rtt = 50.0, loss = 0.0))
            .plusSample(sample(rtt = 350.0, loss = 6.0))

        @Suppress("UNCHECKED_CAST")
        val distribution = fieldsOf(analytics)["qualityDistribution"] as Map<String, Double>
        assertThat(distribution.keys).containsExactly("excellent", "good", "fair", "poor")
        assertThat(distribution["excellent"]).isWithin(1e-9).of(2.0 / 3.0)
        assertThat(distribution["poor"]).isWithin(1e-9).of(1.0 / 3.0)
        assertThat(distribution["good"]).isEqualTo(0.0)
        assertThat(distribution["fair"]).isEqualTo(0.0)
    }

    @Test
    fun `an empty call reports an all-zero distribution with every tier present`() {
        @Suppress("UNCHECKED_CAST")
        val distribution = fieldsOf(CallAnalytics(startedAtMs = 0))["qualityDistribution"] as Map<String, Double>

        assertThat(distribution.keys).containsExactly("excellent", "good", "fair", "poor")
        assertThat(distribution.values.sum()).isEqualTo(0.0)
    }

    // --- wire payload completeness ---------------------------------------------

    @Test
    fun `the payload carries every schema-required field`() {
        val fields = fieldsOf(
            CallAnalytics(startedAtMs = 0),
            durationSeconds = 42,
            isVideo = true,
            endReason = CallEndReason.Remote,
        )

        assertThat(fields.keys).containsExactly(
            "setupTimeMs", "durationSeconds", "reconnectionCount", "networkTransitions",
            "averageRtt", "averagePacketLoss", "maxPacketLoss", "codec", "effectsUsed",
            "filtersUsed", "transcriptionUsed", "qualityDistribution", "platform",
            "deviceModel", "isVideo", "endReason",
        )
        assertThat(fields["durationSeconds"]).isEqualTo(42L)
        assertThat(fields["platform"]).isEqualTo("android")
        assertThat(fields["deviceModel"]).isEqualTo("Pixel 8")
        assertThat(fields["isVideo"]).isEqualTo(true)
        assertThat(fields["endReason"]).isEqualTo("remote")
        assertThat(fields["codec"]).isEqualTo("unknown")
        assertThat(fields["networkTransitions"]).isEqualTo(0)
    }

    // --- endReason wire tokens ---------------------------------------------------

    @Test
    fun `every end reason maps to a stable wire token`() {
        assertThat(CallEndReason.Local.wireValue).isEqualTo("local")
        assertThat(CallEndReason.Remote.wireValue).isEqualTo("remote")
        assertThat(CallEndReason.Rejected.wireValue).isEqualTo("rejected")
        assertThat(CallEndReason.Missed.wireValue).isEqualTo("missed")
        assertThat(CallEndReason.ConnectionLost.wireValue).isEqualTo("connectionLost")
        assertThat(CallEndReason.Failed("boom").wireValue).isEqualTo("failed")
    }
}
